File: packages/depz-sensor-sdk-ts/src/protocol/framing.ts
/**
 * Packet framing and incremental parser (contracts/01_TRANSPORT_FRAMING.md).
 * Mirrors the Python reference (`depz_sensor_sdk.transport.framing`) and the
 * firmware `common/transport/transport.c`. Empty payloads never carry CRC
 * bytes even when crc_type bits are set (contracts/ERRATA.md E6).
 */

import { crc8Maxim, crc16Modbus, crc32IsoHdlc } from "./crc.js";

export const MAGIC0 = 0xa5;
export const MAGIC1 = 0xc3;
export const HEADER_SIZE = 7;
export const MAX_PAYLOAD = 0x3fff;

export enum CrcType {
  None = 0,
  Crc8 = 1,
  Crc16 = 2,
  Crc32 = 3,
}

const CRC_SIZES: Record<CrcType, number> = {
  [CrcType.None]: 0,
  [CrcType.Crc8]: 1,
  [CrcType.Crc16]: 2,
  [CrcType.Crc32]: 4,
};

/** CRC trailer for a payload; empty payloads never carry CRC bytes. */
export function payloadCrcBytes(crcType: CrcType, payload: Uint8Array): Uint8Array {
  if (crcType === CrcType.None || payload.length === 0) return new Uint8Array(0);
  const out = new Uint8Array(CRC_SIZES[crcType]);
  const dv = new DataView(out.buffer);
  if (crcType === CrcType.Crc8) dv.setUint8(0, crc8Maxim(payload));
  else if (crcType === CrcType.Crc16) dv.setUint16(0, crc16Modbus(payload), true);
  else dv.setUint32(0, crc32IsoHdlc(payload), true);
  return out;
}

/**
 * Frame one packet. `crcType` bits are set in the header even for an empty
 * payload (matching device TX), but CRC bytes are only appended for
 * non-empty payloads.
 */
export function buildPacket(
  cmd: number,
  payload: Uint8Array = new Uint8Array(0),
  seq = 0,
  crcType: CrcType = CrcType.None,
): Uint8Array {
  if (payload.length > MAX_PAYLOAD) {
    throw new RangeError(`payload too long: ${payload.length} > ${MAX_PAYLOAD}`);
  }
  const trailer = payloadCrcBytes(crcType, payload);
  const out = new Uint8Array(HEADER_SIZE + payload.length + trailer.length);
  const dataSize = payload.length | (crcType << 14);
  out[0] = MAGIC0;
  out[1] = MAGIC1;
  out[2] = dataSize & 0xff;
  out[3] = dataSize >>> 8;
  out[4] = cmd & 0xff;
  out[5] = seq & 0xff;
  out[6] = crc8Maxim(out.subarray(2, 6));
  out.set(payload, HEADER_SIZE);
  out.set(trailer, HEADER_SIZE + payload.length);
  return out;
}

export interface PacketEvent {
  type: "packet";
  cmd: number;
  seq: number;
  payload: Uint8Array;
}

/**
 * Bytes discarded while hunting for a valid frame. Boundaries between
 * consecutive trash events depend on read chunking; only the concatenated
 * byte stream is deterministic.
 */
export interface TrashEvent {
  type: "trash";
  data: Uint8Array;
}

/** A frame with a valid header whose payload CRC failed; dropped. */
export interface CrcErrorEvent {
  type: "crcError";
  cmd: number;
  seq: number;
}

export type ParserEvent = PacketEvent | TrashEvent | CrcErrorEvent;

/**
 * Incremental frame parser. Feed arbitrary byte chunks; get events.
 * Event order is invariant to chunking (contract 01 §5) except trash event
 * boundaries — concatenate trash data when comparing streams.
 */
export class PacketParser {
  private buf = new Uint8Array(0);
  packets = 0;
  crcErrors = 0;
  headerErrors = 0;
  trashBytes = 0;

  feed(data: Uint8Array): ParserEvent[] {
    if (data.length > 0) {
      const next = new Uint8Array(this.buf.length + data.length);
      next.set(this.buf, 0);
      next.set(data, this.buf.length);
      this.buf = next;
    }
    const out: ParserEvent[] = [];
    for (;;) {
      const ev = this.parseOne();
      if (ev === null) break;
      out.push(ev);
    }
    return out;
  }

  /** Unconsumed bytes currently buffered (diagnostics/tests). */
  get residue(): Uint8Array {
    return this.buf;
  }

  private findMagic(from = 0): number {
    const b = this.buf;
    for (let i = from; i + 1 < b.length; i++) {
      if (b[i] === MAGIC0 && b[i + 1] === MAGIC1) return i;
    }
    return -1;
  }

  private emitTrash(count: number): TrashEvent {
    const data = this.buf.slice(0, count);
    this.buf = this.buf.slice(count);
    this.trashBytes += count;
    return { type: "trash", data };
  }

  private parseOne(): ParserEvent | null {
    const buf = this.buf;
    const pos = this.findMagic();
    if (pos === -1) {
      // Keep the last byte: it may be a split 0xA5.
      if (buf.length > 1) return this.emitTrash(buf.length - 1);
      return null;
    }
    if (pos > 0) return this.emitTrash(pos);
    if (buf.length < HEADER_SIZE) return null;
    const dataSize = buf[2]! | (buf[3]! << 8);
    const payloadSize = dataSize & MAX_PAYLOAD;
    const crcType = ((dataSize >>> 14) & 0x03) as CrcType;
    if (crc8Maxim(buf.subarray(2, 6)) !== buf[6]) {
      // Header corrupt: advance one byte past the magic start and let the
      // magic hunt resync (firmware: rb_skip(off + 1)).
      this.headerErrors += 1;
      return this.emitTrash(1);
    }
    // ERRATA E6: empty payload never carries CRC bytes.
    const crcSize = payloadSize > 0 ? CRC_SIZES[crcType] : 0;
    const total = HEADER_SIZE + payloadSize + crcSize;
    if (buf.length < total) return null;
    const cmd = buf[4]!;
    const seq = buf[5]!;
    const payload = buf.slice(HEADER_SIZE, HEADER_SIZE + payloadSize);
    const trailer = buf.subarray(HEADER_SIZE + payloadSize, total);
    this.buf = this.buf.slice(total);
    if (crcSize > 0) {
      const expected = payloadCrcBytes(crcType, payload);
      for (let i = 0; i < crcSize; i++) {
        if (expected[i] !== trailer[i]) {
          this.crcErrors += 1;
          return { type: "crcError", cmd, seq };
        }
      }
    }
    this.packets += 1;
    return { type: "packet", cmd, seq, payload };
  }
}

File: packages/depz-sensor-sdk-python/depz_sensor_sdk/transport/framing.py
"""Packet framing and incremental parser (contracts/01_TRANSPORT_FRAMING.md).

Byte-exact with firmware `common/transport/transport.c`. The parser fixes the
reference host tool's empty-payload sizing bug (contracts/ERRATA.md E6): a
packet whose header advertises a payload CRC type but carries an empty
payload has **no** CRC bytes on the wire.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass
from enum import IntEnum

from .crc import crc8_maxim, crc16_modbus, crc32_iso_hdlc

MAGIC = b"\xa5\xc3"
HEADER_SIZE = 7
MAX_PAYLOAD = 0x3FFF


class CrcType(IntEnum):
    NONE = 0
    CRC8 = 1
    CRC16 = 2
    CRC32 = 3


_CRC_SIZES = {CrcType.NONE: 0, CrcType.CRC8: 1, CrcType.CRC16: 2, CrcType.CRC32: 4}


def payload_crc_bytes(crc_type: CrcType, payload: bytes) -> bytes:
    """CRC trailer for a payload; empty payloads never carry CRC bytes."""
    if crc_type == CrcType.NONE or not payload:
        return b""
    if crc_type == CrcType.CRC8:
        return struct.pack("<B", crc8_maxim(payload))
    if crc_type == CrcType.CRC16:
        return struct.pack("<H", crc16_modbus(payload))
    return struct.pack("<I", crc32_iso_hdlc(payload))


def build_packet(
    cmd: int, payload: bytes = b"", seq: int = 0, crc_type: CrcType = CrcType.NONE
) -> bytes:
    """Frame one packet. `crc_type` bits are set in the header even for an
    empty payload (matching device TX), but CRC bytes are only appended for
    non-empty payloads."""
    if len(payload) > MAX_PAYLOAD:
        raise ValueError(f"payload too long: {len(payload)} > {MAX_PAYLOAD}")
    data_size = len(payload) | (int(crc_type) << 14)
    ds = struct.pack("<H", data_size)
    hdr_crc = crc8_maxim(ds + bytes((cmd, seq & 0xFF)))
    return MAGIC + ds + bytes((cmd, seq & 0xFF, hdr_crc)) + payload + payload_crc_bytes(
        crc_type, payload
    )


@dataclass(frozen=True)
class Packet:
    cmd: int
    seq: int
    payload: bytes


@dataclass(frozen=True)
class Trash:
    """Bytes discarded while hunting for a valid frame. Boundaries between
    consecutive Trash events depend on read chunking; only the concatenated
    byte stream is deterministic."""

    data: bytes


@dataclass(frozen=True)
class CrcError:
    """A frame with a valid header whose payload CRC failed; dropped."""

    cmd: int
    seq: int


ParserEvent = Packet | Trash | CrcError


class PacketParser:
    """Incremental frame parser. Feed arbitrary byte chunks; get events.

    Event order is invariant to chunking (contract 01 §5) except Trash event
    boundaries — concatenate Trash data when comparing streams.
    """

    def __init__(self) -> None:
        self._buf = bytearray()
        self.packets = 0
        self.crc_errors = 0
        self.header_errors = 0
        self.trash_bytes = 0

    def feed(self, data: bytes) -> list[ParserEvent]:
        self._buf.extend(data)
        out: list[ParserEvent] = []
        while True:
            ev = self._parse_one()
            if ev is None:
                break
            out.append(ev)
        return out

    def _emit_trash(self, count: int) -> Trash:
        data = bytes(self._buf[:count])
        del self._buf[:count]
        self.trash_bytes += count
        return Trash(data)

    def _parse_one(self) -> ParserEvent | None:
        buf = self._buf
        pos = buf.find(MAGIC)
        if pos == -1:
            # Keep the last byte: it may be a split 0xA5.
            if len(buf) > 1:
                return self._emit_trash(len(buf) - 1)
            return None
        if pos > 0:
            return self._emit_trash(pos)
        if len(buf) < HEADER_SIZE:
            return None
        data_size = struct.unpack_from("<H", buf, 2)[0]
        payload_size = data_size & MAX_PAYLOAD
        crc_type = CrcType((data_size >> 14) & 0x03)
        if crc8_maxim(bytes(buf[2:6])) != buf[6]:
            # Header corrupt: advance one byte past the magic start and let
            # the magic hunt resync (firmware: rb_skip(off + 1)).
            self.header_errors += 1
            return self._emit_trash(1)
        # ERRATA E6: empty payload never carries CRC bytes.
        crc_size = _CRC_SIZES[crc_type] if payload_size > 0 else 0
        total = HEADER_SIZE + payload_size + crc_size
        if len(buf) < total:
            return None
        cmd, seq = buf[4], buf[5]
        payload = bytes(buf[HEADER_SIZE : HEADER_SIZE + payload_size])
        trailer = bytes(buf[HEADER_SIZE + payload_size : total])
        del buf[:total]
        if crc_size and payload_crc_bytes(crc_type, payload) != trailer:
            self.crc_errors += 1
            return CrcError(cmd, seq)
        self.packets += 1
        return Packet(cmd, seq, payload)
